File: Algorithms/CPP/606 Construct String from Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    string tree2str(TreeNode* t) {
        return preorder(t);
    }

    string preorder(TreeNode* node) {
        if (node == NULL) {
            return "";
        }

        string left = preorder(node->left);
        string right = preorder(node->right);
        string ans = to_string(node->val);

        if (!right.empty()) {
            ans += "(" + left + ")" + "(" + right + ")";
        }
        else if (!left.empty()) {
            ans += "(" + left + ")";
        }

        return ans;
    }
};

File: Algorithms/Python/606 Construct String from Binary Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def tree2str(self, t: TreeNode) -> str:
        def preorder(node):
            if not node:
                return None
            
            left = preorder(node.left)
            right = preorder(node.right)

            ans = str(node.val)

            if right:
                if left:
                    ans += '(' + left + ')' + '(' + right + ')'
                else:
                    ans += '()' + '(' + right + ')'
            else:
                if left:
                    ans += '(' + left + ')'

            return ans

        ans = preorder(t)
        return ans if ans else ''
